From abdbde84c3f31a6f498bb7afcfeba0e2a15c5801 Mon Sep 17 00:00:00 2001
From: Artem Lapkin <art@khadas.com>
Date: Thu, 12 Aug 2021 12:21:20 +0800
Subject: [PATCH] arm64: dts: meson: khadas-vim3: remake simple sound for i2s
 hdmi spdif output

Remake sound device to i2s hdmi spdif output

Signed-off-by: Artem Lapkin <art@khadas.com>
---
 .../boot/dts/amlogic/meson-khadas-vim3.dtsi   | 79 +++++++++++--------
 1 file changed, 47 insertions(+), 32 deletions(-)

diff --git a/arch/arm64/boot/dts/amlogic/meson-khadas-vim3.dtsi b/arch/arm64/boot/dts/amlogic/meson-khadas-vim3.dtsi
index 3cf4ecb6d..63c384f7c 100644
--- a/arch/arm64/boot/dts/amlogic/meson-khadas-vim3.dtsi
+++ b/arch/arm64/boot/dts/amlogic/meson-khadas-vim3.dtsi
@@ -167,20 +167,20 @@ hdmi_connector_in: endpoint {
 		};
 	};
 
+	spdif_dit: audio-codec-1 {
+		#sound-dai-cells = <0>;
+		compatible = "linux,spdif-dit";
+		status = "okay";
+		sound-name-prefix = "DIT";
+	};
 
 	sound {
 		compatible = "amlogic,axg-sound-card";
 		model = "KHADAS-VIM3";
-		audio-aux-devs = <&tdmin_a>, <&tdmout_a>;
-		audio-routing = "TDMOUT_A IN 0", "FRDDR_A OUT 0",
-				"TDMOUT_A IN 1", "FRDDR_B OUT 0",
-				"TDMOUT_A IN 2", "FRDDR_C OUT 0",
-				"TDM_A Playback", "TDMOUT_A OUT",
-				"TDMIN_A IN 0", "TDM_A Capture",
-				"TDMIN_A IN 3", "TDM_A Loopback",
-				"TODDR_A IN 0", "TDMIN_A OUT",
-				"TODDR_B IN 0", "TDMIN_A OUT",
-				"TODDR_C IN 0", "TDMIN_A OUT";
+		audio-aux-devs = <&tdmout_b>;
+		audio-routing = "TDMOUT_B IN 0", "FRDDR_A OUT 1",
+				"TDM_B Playback", "TDMOUT_B OUT",
+				"SPDIFOUT IN 0", "FRDDR_A OUT 3";
 
 		assigned-clocks = <&clkc CLKID_MPLL2>,
 				  <&clkc CLKID_MPLL0>,
@@ -195,29 +195,9 @@ dai-link-0 {
 			sound-dai = <&frddr_a>;
 		};
 
-		dai-link-1 {
-			sound-dai = <&frddr_b>;
-		};
-
-		dai-link-2 {
-			sound-dai = <&frddr_c>;
-		};
-
-		dai-link-3 {
-			sound-dai = <&toddr_a>;
-		};
-
-		dai-link-4 {
-			sound-dai = <&toddr_b>;
-		};
-
-		dai-link-5 {
-			sound-dai = <&toddr_c>;
-		};
-
 		/* 8ch hdmi interface */
 		dai-link-6 {
-			sound-dai = <&tdmif_a>;
+			sound-dai = <&tdmif_b>;
 			dai-format = "i2s";
 			dai-tdm-slot-tx-mask-0 = <1 1>;
 			dai-tdm-slot-tx-mask-1 = <1 1>;
@@ -226,7 +206,7 @@ dai-link-6 {
 			mclk-fs = <256>;
 
 			codec {
-				sound-dai = <&tohdmitx TOHDMITX_I2S_IN_A>;
+				sound-dai = <&tohdmitx TOHDMITX_I2S_IN_B>;
 			};
 		};
 
@@ -238,6 +218,16 @@ codec {
 				sound-dai = <&hdmi_tx>;
 			};
 		};
+
+		/* spdif or toslink interface */
+		dai-link-4 {
+			sound-dai = <&spdifout>;
+
+			codec-0 {
+				sound-dai = <&spdif_dit>;
+			};
+		};
+
 	};
 
 	wifi32k: wifi32k {
@@ -256,6 +246,29 @@ &clkc_audio {
 	status = "okay";
 };
 
+&tdmif_b {
+	assigned-clocks = <&clkc_audio AUD_CLKID_TDM_SCLK_PAD1>,
+			  <&clkc_audio AUD_CLKID_TDM_LRCLK_PAD1>;
+	assigned-clock-parents = <&clkc_audio AUD_CLKID_MST_B_SCLK>,
+				 <&clkc_audio AUD_CLKID_MST_B_LRCLK>;
+	assigned-clock-rates = <0>, <0>;
+
+	pinctrl-0 = <&tdm_b_fs_pins>, <&tdm_b_dout0_pins>,
+		    <&tdm_b_sclk_pins>, <&mclk0_a_pins>;
+	pinctrl-names = "default";
+	status = "okay";
+};
+
+&tdmout_b {
+    status = "okay";
+};
+
+&spdifout {
+    pinctrl-0 = <&spdif_ao_out_pins>;
+    pinctrl-names = "default";
+    status = "okay";
+};
+
 &cec_AO {
 	pinctrl-0 = <&cec_ao_a_h_pins>;
 	pinctrl-names = "default";
@@ -467,6 +480,7 @@ w25q128: spi-flash@0 {
 	};
 };
 
+/*
 &tdmif_a {
 	status = "okay";
 };
@@ -490,6 +504,7 @@ &toddr_b {
 &toddr_c {
 	status = "okay";
 };
+*/
 
 &tohdmitx {
 	status = "okay";
-- 
2.25.1

